### ISSUE TITLE
Unable to do storylines

### ISSUE THREAD

**Issue by Pluckyduck** [ISeeDEDPpl/Questor#187]
Exception during event execution: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.

   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)

   at Questor.Modules.Caching.Cache.CheckifRouteIsAllHighSec() in C:\Program Files (x86)\InnerSpace.NET Programs\Questor.Modules\Caching\Cache.cs:line 1996

   at Questor.Storylines.Storyline.GotoAgent(StorylineState nextState) in C:\Program Files (x86)\InnerSpace.NET Programs\Questor\Storylines\Storyline.cs:line 192

   at Questor.Storylines.Storyline.ProcessState() in C:\Program Files (x86)\InnerSpace.NET Programs\Questor\Storylines\Storyline.cs:line 271

   at Questor.Behaviors.CombatMissionsBehavior.ProcessState() in C:\Program Files (x86)\InnerSpace.NET Programs\Questor\Behaviors\CombatMissionsBehavior.cs:line 1326

   at Questor.Questor.OnFrame(Object sender, EventArgs e) in C:\Program Files (x86)\InnerSpace.NET Programs\Questor\Questor.cs:line 537

   at ..(DirectEve , Object , EventArgs )

   at DirectEve.InnerSpaceFramework.FrameHook(Object sender, EventArgs e)

   at LavishScriptAPI.LavishScript.Events.Event.Target(Int32 argc, UInt32 argvToken, Int32 ThisObject)

**Comment by Pluckyduck** [comment]
so i installed VS2010 on the  computer and updated .NET.  I compiled with the compile.bat and got the same error.   At this point i open the mission details and right clicked and set destination on the station to talk to agent and nothing happened.  I set destination for the solar system and it instantly took off and started heading in that direction.

**Comment by Escalati0n** [comment]
I have the same error with storylines

**Comment by chantling** [comment]
This is an issue introduced in commit ef58b5de2e949579a750ddd6333a0ec5df687ae9 - "Account for eve client processing time when calculating the route and...".  Storyline.cs sets the storyline agent's destination for Questor's traveler function, but doesn't set it in the Eve autopilot; the traveler does that after the new check for highsec routes.  So when the new CheckifRouteIsAllHighSec() function in Cache.cs calls GetDestinationPath() and tries to assign it to currentPath, there's no autopilot route to assign.  You could try to get around the exception using TryGetValue(), but because there's no autopilot path to get, the function as written will simply cycle endlessly, waiting for a path that'll never appear (I think the idea here was to cycle until the path does appear, giving the client some time to catch up to Questor, but as far as I've been able to tell, you'd never reach a point where the path is generated b/c of the new checks below this call.

A quick and dirty workaround (my favorite - ok, only kind I'm capable of) is to create a small function in Traveler.cs which creates an autopilot route to a given station and call that function before checking for lowsec routes.  Keep in mind that the security level in CheckifRouteIsAllHighSec() needs to be changed from 0.5 to 0.45 or you'll get false low sec alerts on systems with a truesec below 5.

  In addition, the GotoAgent() function gets called continuously while traveler is heading to the agent's station; this means that when you do arrive in the system and the autopilot route is deleted, you'll get the above error again.  I set a flag that only goes through the whole destination setting/checking process once; the flag is cleared when we reach our destination.

It is possible that I misread the new additions in the GotoAgent() function and simply fixing the exception will allow everything to function as intended.  Hope so.  My hack is ugly, ugly.

**Comment by Pluckyduck** [comment]
he had pushed this to help solve the problem that it would not work for my storylines.  I could manually add the station route and it would not go but i could manually add the route to the mission location and it would instantly continue. The route was all highsec but i dont know if the route fell with in the low sec alert limits.

**Comment by chantling** [comment]
Did the commit he pushed fix it for you?  When did you first start seeing an issue?  Because I was running storylines with no issues until I merged this commit (and possibly one or two older) this morning - next storyline I got, I got the same crash as you.  Tracked it back to the issue I described above, whipped out the duct tape and bandaids, and I've run two storylines since with no issues.

TBH, I didn't do much testing to see if it was all storylines or just Materials for War, so I don't know for sure the extent of the issue.  It may have only affected courier missions; didn't track that far back.

**Comment by Pluckyduck** [comment]
no the push did not fix mine.  I rolled back before the repair function was installed to get mine working.  That was the last working build that i had. It did not work on more than just the Mat for War.  I cant remember the other story line that i test it on thought.

**Comment by Pluckyduck** [comment]
Chantiling, thanks for the  .5 vs .45 setting, one of the missions was flagging up as a lowsec system and this solved part of it.   I do not have the last branch working, but i have an older one that i have added some of the support functions to.

**Comment by bbday** [comment]
the problem is function if (!Cache.Instance.CheckifRouteIsAllHighSec()) return; for now you must comment this function inside storylines.

I tried to rewrote the function with this but we have a problem inside DE for me don't working function (trytoget):

private DirectSolarSystem dest = null;

public bool CheckifRouteIsAllHighSec()
        {
            Cache.Instance.RouteIsAllHighSecBool = false;
            // Find the first waypoint
            List<long> currentPath = DirectEve.Navigation.GetDestinationPath();

```
        if (currentPath == null || !currentPath.Any()) return false;

        for (int i = currentPath.Count - 1; i >= 0; i--)
        {
            if (Cache.Instance.dest != null)
                Logging.Log("test", "dest " + dest.Name, Logging.magenta);

            if (DirectEve.SolarSystems.TryGetValue(currentPath[i], out Cache.Instance.dest))
            {
                Logging.Log("test", "dest " + Cache.Instance.dest.Name, Logging.magenta);
            }
            else
            {
                Logging.Log("test", "false trytoget long:" + currentPath[i], Logging.magenta);
                return false;
            }

            if (Cache.Instance.dest.Security < 0.5)
            {
                Logging.Log("CheckifRouteIsAllHighSec", "low sec", Logging.orange);
                Cache.Instance.RouteIsAllHighSecBool = false;
                return true;
            }
            else
                Logging.Log("CheckifRouteIsAllHighSec", "hight sec", Logging.green);
        }
        Cache.Instance.RouteIsAllHighSecBool = true;
        return true;
    }
```

**Comment by chantling** [comment]
I thought my solution written above was working; ran 5 storylines successfully and then last night I had one throw the exceptions.  It was giving me hell to track down b/c it was difficult to reproduce, but I finally tracked it back to an error in my code that only ran the fix the first time Questor went down that code path.  So now I'm back to testing to see if it fixed it right this time or not.  Unfortunately, it'll be a bit of time before I know; I don't normally run storylines because of standings, so I had a bunch of them saved before.  Ran through them all testing the first fix.

**Comment by ISeeDEDPpl** [comment]
chantling's fix actually seems the most reasonable... 
I'll hope that when he has it working (or before?) he'll post it somewhere or commit it to his tree in github.

I really really need to stop assuming that anything in ahaw's github tree has been play tested =/

**Comment by chantling** [comment]
Well, it appears to be working.  Can't say it's been exhaustively tested; 4 courier missions and 5 Materials for War Preparation.  Didn't get a single combat storyline for some reason, but they follow the same code path as couriers, so _theoretically_ they should be good too.  I wouldn't leave it alone until you've run a couple.

Of course, I still have no idea if ahaw's reactivated high sec check, the cause of all of this, actually works.  None of the storyline agents I'm anywhere near are in low sec; even when I've accumulated 5 or 6 storylines, I don't recall ever seeing a low sec agent in there.

No point in updating my tree; I lost interest in Eve for a year and when I came back, my tree was hopelessly deprecated.  I just shamelessly stole yours, ISeeDEDPpl =/  I can email you the files changed (just Cache.cs, Traveler.cs, and Storyline.cs...I hope...) if you like.
